src/lib/optimizer/strategy/chunk_pruning_rule.cpp

```diff
@@ -239,4 +242,107 @@ std::shared_ptr<TableStatistics> ChunkPruningRule::_prune_table_statistics(const
                                            old_statistics.row_count - static_cast<float>(num_rows_pruned));
 }
 
+std::vector<PredicateChain> ChunkPruningRule::_find_predicate_chains_recursively(
+    const std::shared_ptr<StoredTableNode>& stored_table_node, const std::shared_ptr<AbstractLQPNode>& node,
+    PredicateChain current_predicate_chain) {
+  std::vector<PredicateChain> predicate_chains;
+
+  visit_lqp_upwards(node, [&](const auto& current_node) {
+    /**
+     * In the following switch-statement, we
+     *  (1) add PredicateNodes to the current predicate chain, if applicable, and
+     *  (2) check whether the predicate chain continues or ends with current_node.
+     */
+    auto predicate_chain_continues = true;
+    switch (current_node->type) {
+      // clang-format off
+      case LQPNodeType::Alias:
+      case LQPNodeType::Sort:
+      case LQPNodeType::StoredTable:
+      case LQPNodeType::Validate:
+      case LQPNodeType::Projection:   break;
+      case LQPNodeType::Predicate: {
+        const auto& predicate_node = std::static_pointer_cast<PredicateNode>(current_node);
+
+        // PredicateNode might not belong to the current predicate chain,
+        // e.g. when it follows a JoinNode and contains LQPColumnExpressions from other StoredTableNodes.
+        auto belongs_to_predicate_chain = true;
+        const auto& predicate_expression = predicate_node->predicate();
+        visit_expression(predicate_expression, [&](const auto& expression) {
+          if (expression->type != ExpressionType::LQPColumn) return ExpressionVisitation::VisitArguments;
+          const auto& column_expression = std::static_pointer_cast<LQPColumnExpression>(expression);
+          if (column_expression->original_node.lock() != stored_table_node) {
+            // PredicateNode filters stored_table_node. Therefore, we want to add it to the current predicate chain.
+            belongs_to_predicate_chain = false;
+          }
+          return ExpressionVisitation::DoNotVisitArguments;
+        });
+        if (belongs_to_predicate_chain) current_predicate_chain.emplace_back(predicate_node);
+      } break;
+      case LQPNodeType::Join: {
+        auto join_node = std::static_pointer_cast<JoinNode>(current_node);
+        if (join_node->join_mode != JoinMode::Semi) break;
```

Comment:
If chains can only continue after semi joins (where no expression from the "other" comes out of the join), why do we need the check above?